@alloy => Migrate to WKWebView

orta opened the pull request:
This still isn't 100% but it's getting there, and greens.

orta commented on Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m:
```diff
@@ -38,38 +38,39 @@ - (BOOL)isContentStale;
+//{
+//    [super webView:webView didFailLoadWithError:error];
+//    self.hasSuccessfullyLoadedLastRequest = NO;
+//}
```
We can't do this checking ( https://github.com/artsy/eigen/commit/afcd147f4ad5 ) WKWebView has its own cache we can't access.

orta replied:
traded the URL check for a look at the progress

alloy replied:
What about `-[WKNavigationDelegate webView:decidePolicyForNavigationResponse:decisionHandler:]`, the `WKNavigationResponse` object has the `NSURLResponse` object.

orta commented:
OK, this is feeling like it can only really get done with some pair programming, @jorystiefel want to do some with me tomorrow?

jorystiefel commented:
@orta thought you'd never ask

orta commented:
one more test, and then some on-device testing needing.

alloy commented:
Groovy!

orta commented:
This should probably green now. I will do some on-device testing.

alloy commented:
And more rebasing

![](http://media2.giphy.com/media/whOs1JywNpe6c/giphy.gif)

orta commented:
OK, rebased, green'd locally and is ready for review/shipping.

orta commented:
Fixes https://github.com/artsy/eigen/issues/288

alloy commented on Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m:
```diff
@@ -29,41 +25,55 @@ - (void)viewDidLoad
+
+- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation;
+{
+    self.webView.scrollView.backgroundColor = [UIColor blackColor];
```
Can this not be set from `viewDidLoad` ?

orta replied:
No, it makes the full screen black, doing it here does it after it has content on it.

alloy replied:
Ah gotcha, so setting it to black earlier would make the whole view go black.

alloy commented on Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m:
```diff
@@ -29,41 +25,55 @@ - (void)viewDidLoad
 - (void)exitOnboarding
 {
-    [self.personalizeDelegate dismissOnboardingWithVoidAnimation:YES];
+    [self.personalizeDelegate didSignUpAndLogin];
```
Is it correct that a user has signed-up and signed-in if this is called from the `webView:didFailNavigation:withError:` callback?

alloy commented on Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.h:
```diff
@@ -1,14 +1,17 @@
-// Initializing without an NSURL also causes problems in tests.
-- (instancetype)init __attribute__((unavailable("Designated Initializer initWithURL: must be used.")));
+// Private
+- (WKNavigationActionPolicy)shouldLoadNavigationAction:(WKNavigationAction *)navigationAction;
```
Is this for testing?

alloy commented on Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m:
```diff
@@ -34,9 +39,17 @@ - (void)viewDidLoad
+    webView.scrollView.delegate = self;
+    webView.scrollView.decelerationRate = UIScrollViewDecelerationRateNormal;
+
+    _webView = webView;
```
Use the dot, Luke!

orta replied:
Hah, but it's a read only property, having it as read-write encourages more write-ness, but if this is the eigen standard - happy to oblige

orta replied:
doing it this way means you can easily look for any setting specifically because it has to be done via the `_x = y` format

alloy replied:
But you _are_ writing to it, ain’t ya? :) Just make it read-write in the implementation only.

You can also look for it easily with `self.x`, so I don’t think that really matters.

orta replied:
:P

alloy commented on Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m:
```diff
@@ -55,37 +68,39 @@ - (void)showAdminDetails:(UILongPressGestureRecognizer *)gesture
 }
 
-- (void)setupConstraints
+- (void)viewWillLayoutSubviews
```
I think this should be `updateViewConstraints`

orta replied:
:+1:

alloy commented on Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m:
```diff
@@ -4,30 +4,17 @@
 #import "ARInternalShareValidator.h"
 #import "ARAppDelegate.h"
 
-@import TPDWeakProxy;
```
Also remove this pod from the Podfile.

orta replied:
:tada:

alloy commented on Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m:
```diff
@@ -52,7 +39,7 @@ - (instancetype)initWithURL:(NSURL *)url
 
     if (![urlString isEqualToString:url.absoluteString]) {
-        ARActionLog(@"Rewriting %@ as %@", urlString, url.absoluteString);
+        NSLog(@"Rewriting %@ as %@", urlString, url.absoluteString);
```
`ARActionLog` isn’t working right anymore, is it?

alloy commented on Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m:
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
+    [self showLoading];
+
+    // KVO on progress for when we can show the page
+    [self.webView addObserver:self forKeyPath:@"estimatedProgress" options:NSKeyValueObservingOptionNew & NSKeyValueObservingOptionOld context:ARProgressContext];
```
You can just pass `0` for the options here, right?

orta replied:
not sure, so I opted for the most obvious

alloy commented on Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m:
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
-    [self reload];
+    if (context == ARProgressContext && [keyPath isEqualToString:@"estimatedProgress"]) {
+        [self.webView evaluateJavaScript:@"document.readyState == \"interactive\"" completionHandler:^(id response, NSError *error) {
+            if ([response respondsToSelector:@selector(boolValue)]) {
```
Could `response` be anything other than `nil` or `NSCFBoolean`? If not, just do `if ([response boolValue] && !self.loaded) {` ?

alloy replied:
And is the extra `!self.loaded` check really necessary? I guess sending the `-hideLoading` message when it’s already hidden isn’t actually bad, no?

orta replied:
I just didn't want to trust the javascript runtime - was being extra cautious. Happy to switch t something simpler.

alloy replied:
If possible I’d prefer the simpler code, because otherwise the next time we need to change something around this it might seem as if there’s more things to consider than there really are.

orta replied:
removing self.loaded, the `ar_removein...` checks if it needs to run.

alloy commented on Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m:
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
+- (void)hideLoading
 {
-    [self.webView loadRequest:[self requestWithURL:self.currentURL]];
+    self.loaded = YES;
```
`self.loaded = YES` should only be set once the page has been fully loaded, as it is used as a check to see if the page was fully loaded when navigating back to the page. If setting it when hiding the progress indicator, it would be set to `YES` when `DOMContentLoaded` is reached but the page not fully loaded.

orta replied:
cool, thanks, will change logic

alloy commented on Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m:
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
-    if ([readyState isEqualToString:@"interactive"]) {
-        [self removeContentLoadStateTimer];
-        [self hideLoading];
+    if (navigationAction.navigationType == WKNavigationTypeOther) {
```
Why is this needed, wouldn’t it just fall through to the final `return` statement?

orta replied:
Hrm, might be that the logic was shuffled and not re-evaulated at the end, will re-look.

alloy commented on Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m:
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
-}
+    if (navigationAction.navigationType == WKNavigationTypeLinkActivated) {
+        if ([self.shareValidator isSocialSharingURL:URL]) {
+            ARWindow *window = ARAppDelegate.sharedInstance.window;
```
Use `self.webView.window` instead. Getting the right windows is going to get increasingly more important with multi-tasking etc, me thinks.

alloy commented on Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m:
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
+    if (navigationAction.navigationType == WKNavigationTypeLinkActivated) {
+        if ([self.shareValidator isSocialSharingURL:URL]) {
+            ARWindow *window = ARAppDelegate.sharedInstance.window;
+            CGPoint lastTouchPointInView = [window convertPoint:window.lastTouchPoint toView:self.view];
```
Btw, I think you can just omit getting the window completely and instead use `-[UIView convertPoint:fromView:]` with `nil` as the `view` argument, which makes it cover from window coordinates.

orta replied:
we get the touch point _from_ the window, so we need the reference

orta replied:
I think I'm keeping it this way, as it relies on behavior that can only come from the app delegate's uiwindow rather than any window

alloy replied:
That’s what the method does, though https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIView_Class/#//apple_ref/occ/instm/UIView/convertPoint:fromView:

Or I must be understanding the code wrong?

orta replied:
Yeah - we do  this: https://github.com/artsy/eigen/blob/master/Artsy/App/ARAppDelegate.m#L424-L435

orta replied:
cause there's no connection between the touch event and the navigation call

alloy replied:
Oohhhh, I see :+1:

alloy commented on Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m:
```diff
@@ -57,18 +57,18 @@ - (void)markRemoteNotificationsAsRead;
 #pragma mark - Overrides
 
-- (instancetype)initWithURL:(NSURL *)url;
+- (instancetype)initWithURL:(NSURL *)URL;
```
YES! I hate those lowercase `url` variables :D

alloy commented on Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m:
```diff
@@ -83,15 +83,9 @@ - (void)viewDidAppear:(BOOL)animated;
-    NSInteger statusCode = httpResponse.statusCode;
-    self.hasSuccessfullyLoadedLastRequest = statusCode >= 200 && statusCode < 300;
-
+    self.hasSuccessfullyLoadedLastRequest = webView.estimatedProgress == 1;
```
Is this really the same as the replaced code? The old code specifically checks for status codes, because e.g. a `5xx` still means that the webview has fully rendered the document, but a remote error occurred. If a remote error occurred we really want the webview to reload the next time.

orta replied:
this might have been a overzealous merge, will look

alloy commented on Artsy_Tests/View_Controller_Tests/Web_Browsing/ARInternalMobileWebViewControllerTests.m:
```diff
@@ -7,10 +7,19 @@
 @interface ARInternalMobileWebViewController (Testing)
 
 @property (nonatomic, strong) ARInternalShareValidator *shareValidator;
+- (NSURLRequest *)requestWithURL:(NSURL *)url;
```
NoooooooOOOOooOoo, damned lowercase `url` variables come back to haunt me once again 😱

alloy commented:
Maybe I missed it, but it appears that you are not stopping webview loading when the view disappears, which is what `TSMiniBrowser` used to do. With many of our web pages being resource heavy, it kinda makes sense to save the user’s bandwidth once they navigate away to another page.

This is also what the `loaded` ivar was for, so that the webview could be reloaded if it never finished loading the first time around and the user navigates back to the webview.

alloy commented:
Intens change, I love it, finally gone with the ugly chrome and way better progress hooks! :clap: 

![](http://media3.giphy.com/media/N6rK6qtNw0brG/giphy.gif)

orta commented:
### TODO

- [ ] url - URL
- [ ] self.view.window
- [ ] status codes
- [ ] navigationAction.navigationType == WKNavigationTypeOther
- [ ] self.webView evaluateJavaScript

orta commented:
I've updated the PR - I've also removed the web view admin stuff. If I really care ,I'll go in and add something like https://github.com/Naituw/WBWebViewConsole

alloy commented on Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m:
```diff
@@ -60,79 +47,52 @@ - (instancetype)initWithURL:(NSURL *)url
-
-    // As we initially show the loading, we don't want this to appear when you do a back or when a modal covers this view.
-    if (!self.loaded) {
-        [self reload];
```
This reloading, when returning to the view and not fully loading previously, is still missing.

alloy commented on Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m:
```diff
@@ -83,12 +85,27 @@ - (void)viewDidAppear:(BOOL)animated;
 {
-    [super webViewDidFinishLoad:webView];
+    [super webView:webView decidePolicyForNavigationAction:navigationAction decisionHandler:decisionHandler];
+    self.initialRequest = self.initialRequest ?: navigationAction.request;
```
The `initialRequest` needs to be set to `nil` when the webview is reloaded, just like `lastRequestLoadedAt` is set to `nil` from `loadURL:`. Otherwise there’s the chance that a second 500 would be considered ok and it won’t get reloaded.

orta replied:
Legit :+1:

orta commented:
force pushed changes requested this morning

alloy commented:
@orta There’s still a bunch of CI errors about `shouldLoadNavigationAction:`, what gives?

orta commented:
laziness, didn't run tests after my changes, will fix after swift at artsy

alloy commented:
Groovy, baby!

orta commented:
alright, all good again

alloy commented:
Sweet! :dancers: 

![](http://media3.giphy.com/media/mednrGYrVFQWY/giphy.gif)
